### PR TITLE
Replace PostLocalSGDOptimizer with a dedicated model averaging component

### PR DESCRIPTION
**Pull request by wayi1**
## What does this PR do?

This is an improvement of https://github.com/PyTorchLightning/pytorch-lightning/pull/8967.

Replace `PostLocalSGDOptimizer` by a dedicated model averaging component that can run after `optimizer_step` in lightning.py.

The previous implementation can cause data races and fail the training for some cases. For example, if the data loading phase also involves allreduce (e.g., for checking if the input data stream has reached the end), and such allreduce can be kicked off during `optimizer.step()`. Therefore, the implementation in this PR can fix such issue.

Proposal: https://github.com/pytorch/pytorch/issues/59699

There are two options to enable post-local SGD by using vanilla PyTorch:

1.  Post-LocalSGD Comm Hook + Periodic Model Averager
https://github.com/pytorch/pytorch/blob/master/torch/distributed/algorithms/model_averaging/averagers.py#L48-L78

This option requires 1-line addition (such as `averager.average_parameters(model.parameters())`) in the training loop.

2. Post-LocalSGD Comm Hook + Post-LocalSGD Optimizer
https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/post_localSGD_optimizer.py#L15-L48

This option does not need any code change in the training loop. I tried this option earlier. However, one limitation is that if there is another phase such as data loading that also runs some communication like allreduce and overlaps with `optimizer_step()`, then this can cause potential data race and fail the training. Since such concurrent phase like data loading is out of the control of model averaging, I didn't choose this implementation.

### Does your PR introduce any breaking changes? If yes, please list them.

None.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Comment by SciPioneer** [comment]
It's ready to review.

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1562,6 +1572,9 @@ def optimizer_step(
         """
         optimizer.step(closure=optimizer_closure)
 
+        if _TORCH_GREATER_EQUAL_1_10 and self._model_averager is not None:
+            self._model_averager.average_parameters(self.model.parameters())
+
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1577]: in this case, what happens if the LightningModule contains multiple modules for training?

Comment (SciPioneer) [reply pytorch_lightning/core/lightning.py:1577]: Wouldn't 'self.model.parameters()' here recursively contain all the child modules? Or are you suggesting that we should still read parameters from `self.optimizers()`?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -321,43 +322,30 @@ def _register_ddp_hooks(self) -> None:
...
-                optimizer_class=optim_class,
-                averager=averager,
-                **optimizer.defaults,
-            )
-            optimizers[x] = post_localSGD_optimizer
-            del optimizer
-        trainer = self.lightning_module.trainer
-        trainer.optimizers = optimizers
-        trainer.convert_to_lightning_optimizers()
+        self.lightning_module._model_averager = averagers.PeriodicModelAverager(
```
[the remark this answers is not in the document]
Comment (SciPioneer) [reply pytorch_lightning/plugins/training_type/ddp.py:346]: Yes, this will be released as a prototype feature in PyTorch 1.10, and we only have docstring for the documentation purpose so far.

See: https://github.com/pytorch/pytorch/search?q=PeriodicModelAverager

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (SciPioneer) [reply pytorch_lightning/core/lightning.py:1577]: Updated similar to `PostLocalSGDOptimizer`, which only average parameters conditionally;
https://github.com/pytorch/pytorch/blob/master/torch/distributed/optim/post_localSGD_optimizer.py?l#L81-L85

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1562,6 +1572,16 @@ def optimizer_step(
         """
         optimizer.step(closure=optimizer_closure)
 
+        # TODO: Refactor the following code out of this file.
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1575]: Personally, I am not a huge fan of the comm hooks leaking within the Plugin. Ideally, there should be their own isolated components.

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1575]: Here is an example: https://github.com/PyTorchLightning/pytorch-lightning/blob/490c40a8be339afd08dcfdcdd658db6b0be4671b/pytorch_lightning/plugins/ddp_comm_hooks/allreduce_invalid.py

Comment (SciPioneer) [reply pytorch_lightning/core/lightning.py:1575]: Agree that the current implementation is not a very elegant way, and I also think that ideally `lightning.py` file should be untouched in this PR.

I had a discussion with @ananthsub yesterday, who would later let someone do some follow-up refactoring. Sorry that I am not familiar with lightning internals, and actually I am leaving Facebook soon, so I really hope that this PR can be unblocked, and some follow-up works including refactoring and experimentations can be done concurrently.

Thanks for the understanding!

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1562,6 +1572,16 @@ def optimizer_step(
...
+        # CAVEAT: If the user overrides this method, then model averaging will be disabled by default.
+        if _TORCH_GREATER_EQUAL_1_10 and self._model_averager is not None:
+            for opt in self.optimizers(use_pl_optimizer=False):
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1578]: this hook is getting called per optimizer. we still loop over all optimizers?

Comment (SciPioneer) [reply pytorch_lightning/core/lightning.py:1578]: Yes, this hook is called per optimizer step, and we loop over all the optimizers. Certainly we can cache all the target parameters somewhere and only run such looping once. However, I don't think we should include such optimization in this PR:

1) In practice, a trainer usually only has a very small number of optimizers, so the cost of looping here should be trivial. We probably don't want to do over-optimization unless if necessary, and even if we want, it's better to do this incrementally in a separate PR.
2) More important, for some use cases the number of parameters that compute gradients and get averaged can change over time. For example, some base layers may be frozen gradually during pre-train. In this case, it's much safer to run such loop every time.

**Comment by wayi1** [comment]
Close this obsolete PR to favor the replacement https://github.com/PyTorchLightning/pytorch-lightning/pull/12378 from @daniellepintz
